### PR TITLE
[BEAM-2482] CodedValueMutationDetector should use the coders structural value

### PR DESCRIPTION
**Pull request by evindj**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by davorbonaci** [comment]
R: @tgroh 

@tgroh, would you mind taking a look here please?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java [hunk 1]
```diff
@@ -148,13 +148,17 @@ private void verifyUnmodifiedThrowingCheckedExceptions() throws CoderException {
...
-      // ArrayList, which will then be re-encoded in a bounded format. So we really do need to
-      // encode-decode-encode retainedObject.
-      if (Arrays.equals(
-          CoderUtils.encodeToByteArray(coder, clonedOriginalObject),
-          CoderUtils.encodeToByteArray(coder, clonedPossiblyModifiedObject))) {
-        return;
-      }
+      // ArrayList, which will then be re-encoded in a bounded format. So we get a structural value
+      // from a coder and used that to check if the byte array[] are the same. The structuralValue()
+      // method of the Coder returns a StructuralByteArray object.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java]: This is not always the case; the Coder can understand that it has an object with a good `equals` method, and return the object directly.

I'm not sure if I agree that this is a desired change. We'll already have compared the final state of the original object to its clone in the original state, and the clone of the original object to the clone of original object in its final state, so we've already done what is appreciably a fast path, and now are checking the most generous possible "equality" in terms of java.

Comment (evindj) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java]: Hi @lukecwik Please I would like to have your point of view here, do not worry about additional changes I'll revert them

Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java]: The jira issue could be worded better but my intent was we should rely only on the structural value comparison and drop all the other checks which are occurring within this method. Contractually the coder is required to return a value with a good equals method (not needing the usage of deep equals or cloning).

**Comment by evindj** [comment]
I am too much trouble trying to revert some of the changes here. I decided to finally just close this PR and open a new one.

**Pull request closed without merging**